File: problems/0394.0_Decode_String.py
'''
Runtime: 24 ms, faster than 95.62% of Python3 online submissions for Decode String.
Memory Usage: 14.2 MB, less than 79.98% of Python3 online submissions for Decode String.
'''
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        cur_num = 0
        cur_string = ''
        for ch in s:
            if ch == '[':
                stack.append(cur_string)
                stack.append(cur_num)
                cur_string = ''
                cur_num = 0
            elif ch == ']':
                n = stack.pop()
                prev_string = stack.pop()
                cur_string = prev_string + cur_string * n
            elif ch.isdigit():
                cur_num = cur_num * 10 + int(ch)
            else:
                cur_string += ch
        
        return cur_string

File: cpp/0394.0_Decode_String.cpp
/*
Runtime: 0 ms, faster than 100.00% of C++ online submissions for Decode String.
Memory Usage: 6.6 MB, less than 53.22% of C++ online submissions for Decode String.
*/

class Solution {
public:
    string nTimesString(string s, int n) {
        string ans = "";
        for (int i = 0; i < n; i++) {
            ans += s;
        }
        return ans;
    }
    
    string decodeString(string s) {
        vector<string> stack;
        string cur_string = "", cur_num = "";
        for (char ch: s) {
            if (ch == '[') {
                stack.push_back(cur_string);
                stack.push_back(cur_num);
                cur_string = "";
                cur_num = "";
            } else if (ch == ']') {
                string n = stack.back(); stack.pop_back();
                string prev_string = stack.back(); stack.pop_back();
                cur_string = prev_string + nTimesString(cur_string, stoi(n));
            } else if (isdigit(ch)) {
                cur_num += ch;
            } else {
                cur_string += ch;
            }
        }
        return cur_string;
    }
};
